Makefile

```diff
@@ -68,78 +76,108 @@ doc:
-DRUN = docker run      \
-       --init          \
-       --tty           \
-       --interactive   \
-       --network host  \
-       --privileged    \
-       --mount type=bind,source=/var/run/docker.sock,target=/var/run/docker.sock \
-       --mount type=bind,source=${PWD},target=/ebash \
-       --workdir /ebash \
-       --rm
-
-define DOCKER_TEST_TEMPLATE
+DRUN = docker run                                 \
+	--env COLUMNS=${COLUMNS}                      \
+	--init                                        \
+	--tty                                         \
+	--mount type=bind,source=${PWD},target=/ebash \
+	--mount type=bind,source=/var/run/docker.sock,target=/var/run/docker.sock \
+	--network host                                \
+	--privileged                                  \
+	--rm                                          \
+	--workdir /ebash                              \
+
+define DOCKER_TEMPLATE
 
 ifeq ($1,gentoo)
-${1}_CONTAINER = gentoo/stage3
+${1}_IMAGE_BASE = gentoo/stage3
 else ifneq (,$(findstring rocky,$1))
-${1}_CONTAINER = rockylinux/$(subst rocky,rockylinux,$2)
+${1}_IMAGE_BASE = rockylinux/$(subst rocky,rockylinux,$(subst -,:,$1))
 else
-${1}_CONTAINER = $2
+${1}_IMAGE_BASE = $(subst -,:,$1)
 endif
 
+${1}_IMAGE = ghcr.io/elibs/ebash-build-$1
+
+.PHONY: dlint-$1
+dlint-$1: docker-$1
+	${DRUN} $${$1_IMAGE} bin/bashlint --failfast=${FAILFAST} --internal --severity=error --filter=${FILTER} --exclude=${EXCLUDE}
+
 .PHONY: dselftest-$1
-dselftest-$1:
-	bin/ebanner "$2 Dependencies (container=$${$1_CONTAINER})"
-	${DRUN} $${$1_CONTAINER} sh -c "EDEBUG=${EDEBUG} install/all && bin/selftest"
+dselftest-$1: docker-$1
+	${DRUN} $${$1_IMAGE} bin/selftest --severity=error
 
 .PHONY: dtest-$1
-dtest-$1:
-	bin/ebanner "$2 Dependencies (container=$${$1_CONTAINER})"
-	${DRUN} $${$1_CONTAINER} sh -c "EDEBUG=${EDEBUG} install/all && \
-        bin/etest \
-            --break                    \
-            --debug="${EDEBUG}"        \
-            --exclude="${EXCLUDE}"     \
-            --filter="${FILTER}"       \
-            --log-dir=.work            \
-            --repeat=${REPEAT}         \
-            --verbose=${V}             \
-            --work-dir=.work/output"
+dtest-$1: docker-$1
+	${DRUN} $${$1_IMAGE} bin/etest \
+		--break                    \
+		--debug="${EDEBUG}"        \
+		--exclude="${EXCLUDE}"     \
+		--failfast="${FAILFAST}"   \
+		--failures="${FAILURES}"   \
+		--filter="${FILTER}"       \
+		--log-dir=.work            \
+		--repeat=${REPEAT}         \
+		--verbose=${V}             \
+		--work-dir=.work/output
 
 .PHONY: dshell-$1
-dshell-$1:
-	bin/ebanner "$2 Dependencies (container=$${$1_CONTAINER})"
-	${DRUN} $${$1_CONTAINER} sh -c "EDEBUG=${EDEBUG} install/all && /bin/bash"
+dshell-$1: docker-$1
+	${DRUN} $${$1_IMAGE} /bin/bash
+
+.PHONY: docker-$1
+docker-$1:
+	bin/ebanner "Building $${$1_IMAGE}" PULL PUSH
+	bin/ebash docker_build                   \
```

Comment:
This uses ebash provided `docker_build` function to build the docker image. The value this brings over vanilla `docker build` is that it basically does `docker pull || docker build` meaning that it will first try to pull and only if that fails does it fallback to doing a full build. This saves a TON of time.